internal/notification/slack/slack.go

```diff
@@ -0,0 +1,184 @@
+package slack
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/saucelabs/saucectl/internal/report"
+
+	"github.com/rs/zerolog/log"
+	"github.com/slack-go/slack"
+
+	"github.com/saucelabs/saucectl/internal/config"
+	"github.com/saucelabs/saucectl/internal/credentials"
+)
+
+// Reporter represents reporter for slack
+type Reporter struct {
+	Channels       []string
+	TestResults    []report.TestResult
+	Framework      string
+	Metadata       config.Metadata
+	TestEnv        string
+	RenderedResult string
+	Config         config.Notifications
+	Service        Service
+	lock           sync.Mutex
+}
+
+// Add adds the TestResult to the reporter. TestResults added this way can then be rendered out by calling Render().
+func (r *Reporter) Add(t report.TestResult) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.TestResults = append(r.TestResults, t)
+}
+
+// Render renders the test results and sends message to the Slack.
+func (r *Reporter) Render() {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	tables := [][]string{}
+	longestName := 0
+	for _, ts := range r.TestResults {
+		if longestName < len(ts.Name) {
+			longestName = len(ts.Name)
+		}
+	}
+
+	passed := true
+	for _, ts := range r.TestResults {
+		tables = append(tables, []string{statusText(ts.Passed), regenerateName(ts.Name, r.getJobURL(ts.Name, ts.URL), longestName),
+			ts.Platform, ts.DeviceName, ts.Browser, ts.Duration.Truncate(1 * time.Second).String()})
+		if !ts.Passed {
+			passed = false
+		}
+	}
+	var res string
+	for _, t := range tables {
+		res = fmt.Sprintf("%s\n%s", res, strings.Join(t, "\t"))
+	}
+
+	r.RenderedResult = res
+
+	r.sendMessage(passed)
+}
+
+// GetRenderedResult returns rendered result.
+func (r *Reporter) GetRenderedResult() string {
+	return r.RenderedResult
+}
+
+// Reset resets the state of the reporter (e.g. remove any previously reported TestResults).
```

Comment:
The doc is lying here :D Either you do implement a reset that behaves according to the doc, or if resetting is actually not desired for this reporter, update the description to reflect this choice 💪

Reply:
done